File: grokkingAlgorithms/javascript/breadthFirstSearch/breadthFirstSearch/breadthFirstSearch.js
const graph = {};
graph.you = ["alice", "bob", "claire"];
graph.bob = ["anuj", "peggy"];
graph.alice = ["peggy"];
graph.claire = ["thom", "jonny"];
graph.anuj = [];
graph.peggy = [];
graph.thom = [];
graph.jonny = [];

/**
 * Determine whether a person is a seller
 * @param {string} name Friend's name
 * @returns {boolean} Result of checking
 */
const personIsSeller = (name) => name[name.length - 1] === "m";

/**
 * Find a mango seller
 * @param {string} name Friend's name
 * @returns {boolean} Search results
 */
const search = (name) => {
    let searchQueue = [...graph[name]]; // Создание новой очереди и добавление элементов в очередь поиска
    const searched = []; // С помощью этого массива вы можете отслеживать, каких людей вы искали раньше

    while (searchQueue.length) { // Пока очередь не пуста
        const person = searchQueue.shift();
        if (searched.indexOf(person) === -1) { // Если элемент в очереди не был проверен раньше
            if (personIsSeller(person)) {
                console.log(`${person} is a mango seller!`);
                return true;
            }
            searchQueue = searchQueue.concat(graph[person]);
            searched.push(person);
        }
    }

    return false;
}

search("you"); // thom is a mango seller!

File: grokkingAlgorithms/breadthFirstSearch/golang/breadth-first_search.go
package main

import "fmt"

var graph = make(map[string][]string) // string - имя друга, []string - друзья этого друга

func main() {
	graph["you"] = []string{"alice", "bob", "claire"}
	graph["bob"] = []string{"anuj", "peggy"}
	graph["alice"] = []string{"peggy"}
	graph["claire"] = []string{"thom", "jonny"}
	graph["anuj"] = []string{}
	graph["peggy"] = []string{}
	graph["thom"] = []string{}
	graph["jonny"] = []string{}

	search("you")
}

// search поиск в ширину. O(n) + O(m) = O(n + m)
func search(name string) bool {
	var search_queue []string                           // Создание новой очереди
	search_queue = append(search_queue, graph[name]...) // Добавление элементов в очередь поиска
	var searched []string                               // С помощью этого массива вы можете отслеживать, каких людей вы искали раньше
	var person string                                   // С помощью этой переменной вы можете извлекать людей из очереди поиска

	for len(search_queue) != 0 { // Пока очередь не пуста
		// Удаление человека из очереди поиска и его сохранение в переменную person
		person = search_queue[0]
		search_queue = search_queue[1:]
		// Если человек еще не был проверен ранее, проверьте, является ли он продавцом манго
		if person_not_in_searched(person, searched) {
			if person_is_seller(person) {
				fmt.Println(person + " is mango seller!")
				return true
			}

			// Если человек не является продавцом манго, добавьте его друзей в очередь поиска
			search_queue = append(search_queue, graph[person]...)
			// Добавьте человека в массив, чтобы отслеживать, кто уже был проверен
			searched = append(searched, person)

		}
	}

	return false
}

// person_is_seller проверяет на продавца манго.
func person_is_seller(name string) bool {
	// Если последняя буква имени человека - 'm', то он продавец манго
	return name[len(name)-1] == 'm'
}

// person_not_in_searched проверяет, не включался ли человек в поиск ранее.
func person_not_in_searched(person string, searched []string) bool {
	for _, n := range searched {
		// Если человек уже был проверен ранее, верните false
		if n == person {
			return false
		}
	}
	return true
}
